Add highlights for OCaml.

undu opened the pull request:
It's missing function usages and arguments, but they are represented by the same type of node in the same list, it wasn't clear to me how to match the first item of the list only, and the others.

For the rest it should be pretty complete but I might have missed some advanced syntax, especially regarding types.

steelsojka commented:
@undu If you want to match the first item in a list you would use the `.` operator. For example:

```scheme
(formal_parameters . (identifier))
```

This would match the first parameter of a list of parameters.

undu commented:
Thanks for the tip, now function usages with module paths paths up to 3 components are highlighted as such.

I'd prefer a way to generalize the highlighting of those for paths with more elements, but module paths with more than 3 nested modules are exceedingly rare.

I've also added some more elements for types, fields and separators; I've also removed a couple highlight groups which were too greedy.

theHamsta commented on queries/ocaml/hightlights.scm:
```diff
@@ -0,0 +1,146 @@
+  "module"
+  "match"
+  "with"
+  "when"
```
Is this a conditional?

undu replied:
Good catch!

theHamsta commented on queries/ocaml/hightlights.scm:
```diff
@@ -0,0 +1,146 @@
+(infix_expression (infix_operator) @operator
+                  (#eq? @operator "|>")
+                  (value_path (value_path (value_path (value_name) @function))))
+
```
This is not ideal. But I don't see a better solution for now since it would require combining two queries. A predicate for this situation would need to be really specialized. Can we extend the has-ancenstor to accept a Lua function as a predicate for the parent? Or a predicate just for that situation?

undu replied:
I've raised an issue to the ocaml syntax, I think the tree structure can be changed to avoid this uglyness. https://github.com/tree-sitter/tree-sitter-ocaml/issues/40

theHamsta commented on queries/ocaml/hightlights.scm:
```diff
@@ -0,0 +1,146 @@
+(module_name) @structure
```
Structure? I dont think we have this highlight (too lazy to check on mobile). Maybe `@namespace`?

TravonteD replied:
We do have a structure highlight `TSStructure`

undu replied:
I could use either, but structure works.

TravonteD commented on queries/ocaml/hightlights.scm:
```diff
@@ -0,0 +1,146 @@
+(type_binding (type_constructor) @type)
+
+(exception_definition (constructor_declaration (constructor_name) @exception))
+(application_expression (value_path (value_name) @_raise)
```
What is the purpose of the `@_raise` query?

undu replied:
`@_raise` is used to capture the function `raise`, which is always followed by a constructor of an exception. This is because `raise` is a normal function (and value) instead of a keyword.

TravonteD commented on queries/ocaml/hightlights.scm:
```diff
@@ -0,0 +1,146 @@
+  (value_name) @function
+  (function_expression))
+
+(application_expression . (value_path (value_name) @function))
```
It looks like you set `(value_path (value_name) @function)` everywhere, but above where it's `@_raise` you can shortcut this by extracting it and having it at the top-level. The highlighter will then default it to `@function` unless a more specific query is used.

undu replied:
The problem is that `(value_path (value_name))` matches any value, function or not. So I have to depend on the outermost context. I've raised the issue https://github.com/tree-sitter/tree-sitter-ocaml/issues/40 to prevent this awkwardness.

TravonteD commented on queries/ocaml/hightlights.scm:
```diff
@@ -0,0 +1,146 @@
+ "~"
+ "?"
+ "|"
+ "+"
```
the plus, minus, star, etc. should probably be highlighted as `@operator`

undu replied:
They are used for type annotations as well, I need to add context for these.

theHamsta replied:
Namespace does not exist yet. Structure was this what nobody knew what it actually is. Now I remember :-)

TravonteD commented on queries/ocaml/hightlights.scm:
```diff
@@ -0,0 +1,146 @@
+
+(let_binding
+  (value_name) @function
+  (parameter))
```
Should this be highlighted as `@parameter`?

undu replied:
I don't think so: only the `value_name` in the `parameter` subtree should be captured as a parameter. This is done elsewhere.

theHamsta commented:
You can also be the maintainer in the README if you want to

vigoux commented on queries/ocaml/hightlights.scm:
```diff
@@ -0,0 +1,147 @@
+  "end"
+  "fun"
+  "function"
+  "try"
```
Agreed on this, @undu is this exception related ?

undu replied:
yes, it's part of the `try ... with <capture exception> -> ...` block

vigoux replied:
Just `@exception`

vigoux commented:
Awesome ! Thanks again for your contribution !
